### PR TITLE
Allow run-time configuration of OpenCyphal node.

### LINKED ISSUE #175: [Feature Request] Allow runtime configuration of the OpenCyphal Node
<a href="https://107-systems.org/"></a>

### ⚡ Feature Request

Currently various parameters such as size of the o1heap memory area, CAN Classic or CAN FD, size of the CAN receive ringbuffer are all hard-coded constants. This should be changed to that the user can configure their `Node` instance to their hearts delight.

### PR DESCRIPTION
**Pull request by aentinger**
This fixes #175.

### CODE REVIEW

File: src/Node.cpp [hunk 1]
```diff
@@ -117,13 +109,13 @@ void Node::processRxQueue()
   }
 }
 
-void Node::processTxQueue()
+void Node::processTxQueue(CanFrameTransmitFunc tx_func)
```
Comment (pavel-kirienko) [inline src/Node.cpp:112]:
```suggestion
void Node::processTxQueue(const CanFrameTransmitFunc tx_func)
```

File: src/O1Heap.h [hunk 2]
```diff
@@ -5,42 +5,44 @@
...
  **************************************************************************************/
 
 #include <o1heap/o1heap.h>
 
+/**************************************************************************************
+ * DEFINE
+ **************************************************************************************/
+
+#define CYPHAL_DECLARE_HEAP(name,size) \
+  static uint8_t name[size] __attribute__ ((aligned (O1HEAP_ALIGNMENT)))
```
Comment (pavel-kirienko) [inline src/O1Heap.h:22]: In modern C++, there are few valid cases for the use of macros; all known to me modern coding standards strongly discourage their usage for anything except header inclusion and inclusion guards (even those are soon to go the way of the `goto` statement as C++20 finally has modules, yay).

```c++
template <std::size_t Size>
struct alignas(O1HEAP_ALIGNMENT) CyphalHeap final : public std::array<std::uint8_t, Size> {};
```

File: src/O1Heap.h [hunk 3]
```diff
@@ -5,42 +5,44 @@
...
  * CLASS DECLARATION
  **************************************************************************************/
 
-template <size_t HEAP_SIZE>
 class O1Heap
 {
 public:
-
-  O1Heap();
+  O1Heap(uint8_t * heap_ptr, size_t const heap_size)
```
Comment (pavel-kirienko) [inline src/O1Heap.h:31]: I think this should be non-copyable and non-movable.

Comment (aentinger) [reply src/O1Heap.h:33]: I'm not sure I fully get what you mean by this. This wrapper class is a private member of `Node`, hence it does not really matter its copyable or moveable as the user does not get to use it.

File: src/O1Heap.h (hunk 2 above)
Comment (aentinger) [reply src/O1Heap.h]: With you on macros, C++20 is not an option as none of the currently supported Arduino cores are supporting C++20 (or even C++17/14 for that matter ... C++11 is a baseline we can take for granted). I'm testing your proposal in the CI build across all cores now.

File: src/O1Heap.h (hunk 3 above)
Comment (aentinger) [reply src/O1Heap.h]: Eliminated the wrapper class altogether in 77aace9bf0de506ce07f7fb5504346de1c45500f.

File: src/O1Heap.h (hunk 2 above)
Comment (aentinger) [reply src/O1Heap.h]: Derived from std::array works across all cores, see [here](https://github.com/107-systems/107-Arduino-Cyphal/actions/runs/3701421788/jobs/6270710791).

File: src/Node.cpp [hunk 4]
```diff
@@ -21,20 +21,20 @@ Node::Node(uint8_t * heap_ptr,
...
            size_t const mtu_bytes)
-: _o1heap_hdl(heap_ptr, heap_size)
+: _o1heap_ins{o1heapInit(heap_ptr, heap_size)}
 , _canard_hdl{canardInit(Node::o1heap_allocate, Node::o1heap_free)}
 , _canard_tx_queue{canardTxInit(tx_queue_capacity, mtu_bytes)}
 , _canard_rx_queue{rx_queue_capacity}
 {
   _canard_hdl.node_id = node_id;
-  _canard_hdl.user_reference = reinterpret_cast<void *>(&_o1heap_hdl);
+  _canard_hdl.user_reference = reinterpret_cast<void *>(_o1heap_ins);
```
Comment (pavel-kirienko) [inline src/Node.cpp:30]:
```suggestion
  _canard_hdl.user_reference = static_cast<void *>(_o1heap_ins);
```

**Review by pavel-kirienko — APPROVED** [review]

**Pull request merged**